File: src/db/index/common/query.cc
// Copyright 2025-present the zvec project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <cstdint>
#include <zvec/db/query.h>
#include <zvec/db/schema.h>
#include "db/common/constants.h"
#include "db/index/common/type_helper.h"

namespace zvec {

Status QueryTarget::validate(const FieldSchema *schema,
                             bool *need_sanitize) const {
  auto opt_view = get_vector_view();
  auto *fc = get_fts_clause();
  auto &field_name = field_name_;
  auto &query_params = query_params_;
  // A "scalar-only filter" query has no vector payload — either the clause
  // is not a VectorClause (e.g., FtsClause) or its fields are all empty.
  bool no_vector_payload = !opt_view || (opt_view->query_vector_.empty() &&
                                         opt_view->sparse_indices_.empty());

  if (schema == nullptr) {
    if (fc != nullptr) {
      // FTS query requires a valid field_name_ that resolves to an FTS field.
      return Status::InvalidArgument(
          "Invalid query: fts requires a valid FTS field, but field[",
          field_name, "] does not exist in the collection");
    }
    if (no_vector_payload) {
      // Scalar-only filter query
      return Status::OK();
    } else {
      // If a query vector was provided, the field must exist as a vector field
      // since we are performing a vector similarity search.
      return Status::InvalidArgument(
          "Invalid query: query vector is provided, but query field[",
          field_name,
          "] does not exist or is not a vector field in the collection");
    }
  }

  // FTS query: field must be an FTS-indexed field.
  if (fc != nullptr) {
    if (schema->index_type() != IndexType::FTS) {
      return Status::InvalidArgument(
          "Invalid query: fts requires an FTS-indexed field, but field[",
          field_name, "] has index type ",
          IndexTypeCodeBook::AsString(schema->index_type()));
    }
    if (query_params && query_params->type() != schema->index_type()) {
      return Status::InvalidArgument(
          "Invalid query: query params type does not match the index type of "
          "FTS field[",
          field_name, "], expected ",
          IndexTypeCodeBook::AsString(schema->index_type()), " but got ",
          IndexTypeCodeBook::AsString(query_params->type()));
    }
    return Status::OK();
  }

  // Schema is non-null from here on: a vector payload is required.
  if (no_vector_payload) {
    return Status::InvalidArgument(
        "Invalid query: missing query clause for field[", field_name, "]");
  }

  auto &query_vector = opt_view->query_vector_;
  auto &query_sparse_indices = opt_view->sparse_indices_;
  auto &query_sparse_values = opt_view->sparse_values_;

  // Vector query
  if (schema->is_dense_vector()) {
    // Validate dimension
    auto dim = schema->dimension();
    switch (schema->data_type()) {
      case DataType::VECTOR_FP16:
        if (dim * sizeof(float16_t) != query_vector.size()) {
          return Status::InvalidArgument(
              "Invalid query: dimension mismatch, expected ", dim, " but got ",
              query_vector.size() / sizeof(float16_t), " (FP16)");
        }
        break;
      case DataType::VECTOR_FP32:
        if (dim * sizeof(float) != query_vector.size()) {
          return Status::InvalidArgument(
              "Invalid query: dimension mismatch, expected ", dim, " but got ",
              query_vector.size() / sizeof(float), " (FP32)");
        }
        break;
      case DataType::VECTOR_FP64:
        if (dim * sizeof(double) != query_vector.size()) {
          return Status::InvalidArgument(
              "Invalid query: dimension mismatch, expected ", dim, " but got ",
              query_vector.size() / sizeof(double), " (FP64)");
        }
        break;
      case DataType::VECTOR_INT8:
        if (dim * sizeof(int8_t) != query_vector.size()) {
          return Status::InvalidArgument(
              "Invalid query: dimension mismatch, expected ", dim, " but got ",
              query_vector.size() / sizeof(int8_t), " (INT8)");
        }
        break;
      case DataType::VECTOR_INT16:
      case DataType::VECTOR_INT4:
      case DataType::VECTOR_BINARY32:
      case DataType::VECTOR_BINARY64:
        return Status::NotSupported(
            "Invalid query: dense vector type of field[", field_name,
            "] is not supported");
      default:
        return Status::InvalidArgument("Invalid query: field[", field_name,
                                       "] is not a dense vector field");
    }
  } else if (schema->is_sparse_vector()) {
    size_t value_byte_size = 0;
    switch (schema->data_type()) {
      case DataType::SPARSE_VECTOR_FP32:
        value_byte_size = sizeof(float);
        break;
      case DataType::SPARSE_VECTOR_FP16:
        value_byte_size = sizeof(float16_t);
        break;
      default:
        return Status::InvalidArgument(
            "Invalid query: sparse vector type of field[", field_name,
            "] is not supported");
    }
    if (query_sparse_indices.size() % sizeof(uint32_t) != 0 ||
        query_sparse_values.size() % value_byte_size != 0 ||
        query_sparse_indices.size() / sizeof(uint32_t) !=
            query_sparse_values.size() / value_byte_size) {
      return Status::InvalidArgument(
          "Invalid query: sparse vector query for field[", field_name,
          "] has mismatched indices and values sizes");
    }
    size_t n_indices = query_sparse_indices.size() / sizeof(uint32_t);
    if (n_indices > kSparseMaxDimSize) {
      return Status::InvalidArgument(
          "Invalid query: too many sparse indices, the maximum allowed is ",
          kSparseMaxDimSize);
    }
    if (n_indices > 1 && need_sanitize) {
      const auto *idx =
          reinterpret_cast<const uint32_t *>(query_sparse_indices.data());
      auto status = need_sanitize_sparse(idx, n_indices);
      if (status == SparseIndicesStatus::kHasDuplicate) {
        return Status::InvalidArgument(
            "Invalid query: sparse vector query for field[", field_name,
            "] contains duplicate indices");
      }
      if (status == SparseIndicesStatus::kNeedSort) {
        *need_sanitize = true;
      }
    }
  } else {
    return Status::InvalidArgument("Invalid query: field[", field_name,
                                   "] is not a vector field");
  }
  // Validate query_params type
  if (query_params && query_params->type() != schema->index_type()) {
    return Status::InvalidArgument(
        "Invalid query: query params type does not match the index type of "
        "vector field[",
        field_name, "], expected ",
        IndexTypeCodeBook::AsString(schema->index_type()), " but got ",
        IndexTypeCodeBook::AsString(query_params->type()));
  }
  if (query_params && query_params->type() == IndexType::IVF_RABITQ) {
    auto ivf_rabitq_params =
        std::dynamic_pointer_cast<IvfRabitqQueryParams>(query_params);
    if (!ivf_rabitq_params) {
      return Status::InvalidArgument(
          "Invalid query: IVF_RABITQ index requires IvfRabitqQueryParams");
    }
    if (ivf_rabitq_params->nprobe() <= 0) {
      return Status::InvalidArgument(
          "Invalid query: IVF_RABITQ nprobe must be greater than 0");
    }
  }
  return Status::OK();
}

Status validate_topk_and_output_fields(
    int topk, const std::optional<std::vector<std::string>> &output_fields) {
  if ((uint32_t)topk > kMaxQueryTopk) {
    return Status::InvalidArgument("Invalid query: topk[", topk,
                                   "] exceeds the maximum allowed value of ",
                                   kMaxQueryTopk);
  }
  if (output_fields.has_value() &&
      output_fields->size() > kMaxOutputFieldSize) {
    return Status::InvalidArgument(
        "Invalid query: too many output fields, the maximum allowed is ",
        kMaxOutputFieldSize);
  }
  return Status::OK();
}

Status SearchQuery::validate(const FieldSchema *schema,
                             bool *need_sanitize) const {
  if (need_sanitize) {
    *need_sanitize = false;
  }
  auto s = validate_topk_and_output_fields(topk_, output_fields_);
  if (!s.ok()) {
    return s;
  }
  return target_.validate(schema, need_sanitize);
}

Status sanitize_sparse_vector(VectorClause &vc, const FieldSchema *schema) {
  if (!schema || !schema->is_sparse_vector()) {
    return Status::OK();
  }
  size_t value_byte_size = 0;
  switch (schema->data_type()) {
    case DataType::SPARSE_VECTOR_FP32:
      value_byte_size = sizeof(float);
      break;
    case DataType::SPARSE_VECTOR_FP16:
      value_byte_size = sizeof(float16_t);
      break;
    default:
      return Status::OK();
  }
  size_t n_indices = vc.sparse_indices_.size() / sizeof(uint32_t);
  if (n_indices <= 1) {
    return Status::OK();
  }
  if (sort_and_find_duplicates(
          reinterpret_cast<uint32_t *>(vc.sparse_indices_.data()),
          vc.sparse_values_.data(), n_indices, value_byte_size)) {
    return Status::InvalidArgument(
        "Invalid query: sparse vector query for field[", schema->name(),
        "] contains duplicate indices");
  }
  return Status::OK();
}

Status sanitize_sparse_vector(QueryTarget &target, const FieldSchema *schema) {
  if (auto *vvc = target.get_vector_view_clause()) {
    target.clause_ = VectorClause{std::string(vvc->query_vector_),
                                  std::string(vvc->sparse_indices_),
                                  std::string(vvc->sparse_values_)};
  }
  return sanitize_sparse_vector(*target.get_vector_clause(), schema);
}

}  // namespace zvec

File: python/zvec/model/param/query.py
# Copyright 2025-present the zvec project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import annotations

import warnings
from dataclasses import dataclass
from typing import Optional, Union

from ...common import VectorType
from . import FtsQueryParam, HnswQueryParam, HnswRabitqQueryParam, IVFQueryParam

__all__ = ["Fts", "Query", "VectorQuery"]


@dataclass(frozen=True)
class Fts:
    """Full-text search query parameters.

    Attributes:
        query_string (Optional[str]): FTS query expression
            (e.g. '+vector -slow "exact phrase"'). Mutually exclusive with match_string.
        match_string (Optional[str]): Natural language match string,
            tokenized and combined using the default operator.
            Mutually exclusive with query_string.
    """

    query_string: Optional[str] = None
    match_string: Optional[str] = None


@dataclass(frozen=True)
class Query:
    """Represents a search query for a specific field in a collection.

    A `Query` can be constructed for either vector search or full-text search,
    but not both simultaneously.

    For vector search, provide `id` or `vector` (and optionally `param`).
    For FTS, provide `fts`.

    Attributes:
        field_name (str): Name of the field to query.
        id (Optional[str], optional): Document ID to fetch vector from. Default is None.
        vector (VectorType, optional): Explicit query vector. Default is None.
        param (Optional[Union[HnswQueryParam, HnswRabitqQueryParam, IVFQueryParam, FtsQueryParam]], optional):
            Index-specific query parameters. Default is None.
        fts (Optional[Fts], optional): Full-text search parameters. Default is None.

    Examples:
        >>> import zvec
        >>> # Query by ID
        >>> q1 = zvec.Query(field_name="embedding", id="doc123")
        >>> # Query by vector
        >>> q2 = zvec.Query(
        ...     field_name="embedding",
        ...     vector=[0.1, 0.2, 0.3],
        ...     param=HnswQueryParam(ef=300)
        ... )
        >>> # FTS query
        >>> q3 = zvec.Query(
        ...     field_name="content",
        ...     fts=Fts(match_string="machine learning")
        ... )
        >>> # FTS query with custom operator
        >>> q4 = zvec.Query(
        ...     field_name="content",
        ...     fts=Fts(match_string="machine learning"),
        ...     param=FtsQueryParam(default_operator="AND")
        ... )
    """

    field_name: str
    id: Optional[str] = None
    vector: VectorType = None
    param: Optional[
        Union[HnswQueryParam, HnswRabitqQueryParam, IVFQueryParam, FtsQueryParam]
    ] = None
    fts: Optional[Fts] = None

    def has_id(self) -> bool:
        """Check if the query is based on a document ID.

        Returns:
            bool: True if `id` is set, False otherwise.
        """
        return self.id is not None

    def has_vector(self) -> bool:
        """Check if the query contains an explicit vector.

        Returns:
            bool: True if `vector` is non-empty, False otherwise.
        """
        return self.vector is not None and len(self.vector) > 0

    def has_fts(self) -> bool:
        """Check if the query contains an FTS (full-text search) condition.

        Returns:
            bool: True if `fts` is set with a query_string or match_string.
        """
        if self.fts is not None:
            return bool(self.fts.query_string) or bool(self.fts.match_string)
        return False

    def _validate(self) -> None:
        if not isinstance(self.field_name, str) or not self.field_name.strip():
            raise ValueError("Field name must be a non-empty string")
        if self.has_id() and self.has_vector():
            raise ValueError("Cannot provide both id and vector")
        if self.has_fts() and (self.has_vector() or self.has_id()):
            raise ValueError(
                "Cannot combine fts with vector search fields (id/vector) in a single Query"
            )
        if self.fts is not None and self.fts.query_string and self.fts.match_string:
            raise ValueError(
                "Cannot provide both query_string and match_string in Fts; "
                "they are mutually exclusive"
            )


class VectorQuery(Query):
    """Deprecated alias for Query. Use Query instead."""

    def __new__(cls, *args, **kwargs):  # noqa : ARG004
        warnings.warn(
            "VectorQuery is deprecated and will be removed in a future version. "
            "Use Query instead.",
            DeprecationWarning,
            stacklevel=2,
        )
        return super().__new__(cls)
